File: Core/Python/binary_node.py
class BinaryNode:
  value = None
  left = None
  right = None
  parent = None

  def __init__(self, value):
    self.value = value

  def set_left(self, node):
    self.left = node
    self.left.parent = self

  def set_right(self, node):
    self.right = node
    self.right.parent = self

  def left_value(self):
    if (self.left != None):
      return str(self.left.value)
    return ""

  def right_value(self):
    if (self.right != None):
      return str(self.right.value)
    return ""
    
  def __str__(self):
    return "[BinaryNode value: " + str(self.value) + ", left: " + self.left_value() + ", right: " + self.right_value() + "]"

File: Core/JavaScript/binary-node.ts
class BinaryNode<TData> {
    left?: BinaryNode<TData>;
    right?: BinaryNode<TData>;
    value: TData;
    height: number;

    constructor(value: TData, left?: BinaryNode<TData>, right?: BinaryNode<TData>) {
        this.value = value;
        this.left = left;
        this.right = right;
        this.height = 0;
    }

    setRight(node?: BinaryNode<TData>) {
        this.right = node;
        this.right!.height = this.height + 1;
    }

    setLeft(node?: BinaryNode<TData>) {
        this.left = node;
        this.left!.height = this.height + 1;
    }

    leftValue(): TData | undefined {
        if (!this.left) return;
        return this.left.value;
    }

    rightValue(): TData | undefined {
        if (!this.right) return;
        return this.right?.value;
    }

    toString(): string {
        return `[BinaryNode value:${this.value}, left:${this.leftValue() ?? ''}, right:${this.rightValue() ?? ''}]`;
    }
}